CODE/Tests/Fake/FakeManager.swift

```diff
@@ -0,0 +1,184 @@
+    public func addEventHandler(_ handler: @escaping (Gateway.Event) -> Void) async {
+        eventHandlers.append(handler)
+    }
+    public func addEventParseFailureHandler(
+        _ handler: @escaping (Error, String) -> Void
+    ) async { }
+    public nonisolated func disconnect() { }
+    
+    var connectionWaiter: CheckedContinuation<(), Never>?
+    public func waitUntilConnected() async {
+        if self.state == .connected {
+            return
+        } else {
+            await withCheckedContinuation {
+                self.connectionWaiter = $0
+            }
+        }
+    }
+    
+    public func send(key: EventKey) throws {
+        let data = testData(for: key.rawValue)!
+        let decoder = JSONDecoder()
+        let event = try decoder.decode(Gateway.Event.self, from: data)
+        for handler in eventHandlers {
+            handler(event)
+        }
+    }
+    
+    public func sendAndAwaitResponse<T>(
+        key: EventKey,
+        as type: T.Type = T.self,
+        file: String = #file,
+        line: UInt = #line
+    ) async throws -> T {
+        try self.send(key: key)
+        let value = await FakeResponseStorage.shared.awaitResponse(at: key.responseEndpoints[0])
+        let unwrapped = try XCTUnwrap(
+            value as? T,
+            "Value '\(value)' can't be cast to '\(_typeName(T.self))'"
+        )
+        return unwrapped
+    }
+}
+
+public actor FakeResponseStorage {
+    
+    private var continuations = [String: CheckedContinuation<Any, Never>]()
+    private var unhandledResponses = [String: Any]()
+    
+    public init() { }
+    public static var shared = FakeResponseStorage()
+    
+    public func awaitResponse(at endpoint: Endpoint) async -> Any {
+        await withCheckedContinuation { continuation in
+            self.expect(at: endpoint, continuation: continuation)
+        }
+    }
+    
+    private func expect(at endpoint: Endpoint, continuation: CheckedContinuation<Any, Never>) {
+        if let response = unhandledResponses[endpoint.urlSuffix] {
+            continuation.resume(returning: response)
+        } else {
+            continuations[endpoint.urlSuffix] = continuation
+            Task {
+                try await Task.sleep(nanoseconds: 3_000_000_000)
+                try Task.checkCancellation()
+                if continuations.removeValue(forKey: endpoint.urlSuffix) != nil {
+                    XCTFail("Penny did not respond in-time at '\(endpoint.urlSuffix)'")
+                    continuation.resume(with: .success(Optional<Never>.none as Any))
+                }
+            }
+        }
+    }
+    
+    func respond(to endpoint: Endpoint, with payload: Any) {
+        if let continuation = continuations.removeValue(forKey: endpoint.urlSuffix) {
+            continuation.resume(returning: payload)
+        } else {
+            unhandledResponses[endpoint.urlSuffix] = payload
+        }
+    }
+}
+
+private struct FakeDiscordClient: DiscordClient {
```

Comment:
It would be nice to split these out into their own types

Reply:
I assume you mean the `send` and `sendAndAwaitResponse ` functions?

Reply:
Sorry I meant files - `FakeDiscordClient` and `FakeResponseStorage` should probably live in their own files

Reply:
good idea :\)